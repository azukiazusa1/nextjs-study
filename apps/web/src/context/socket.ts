import { createContext } from 'react';
import io, { Socket } from 'socket.io-client';

import config from '@/config';

export const socket = io(config.SOCKET_URL + '/session');
export const SocketContext = createContext<Socket>(undefined as never);
