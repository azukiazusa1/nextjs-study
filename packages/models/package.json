{
  "name": "models",
  "version": "0.0.0",
  "main": "./build/index.js",
  "types": "./build/index.d.ts",
  "license": "MIT",
  "scripts": {
    "lint": "eslint --fix src/**/*.ts",
    "dev": "tsc -w",
    "build": "tsc"
  },
  "dependencies": {
    "eslint-config-custom": "*",
    "eslint": "^8.15.0",
    "typescript": "^4.6.4"
  }
}
