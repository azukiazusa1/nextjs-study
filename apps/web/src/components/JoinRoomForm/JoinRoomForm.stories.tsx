import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import { SocketContext } from '@/context/socket';

import JoinRoomForm from './JoinRoomForm';

export default {
  component: JoinRoomForm,
} as ComponentMeta<typeof JoinRoomForm>;

const dummySocket = {
  emit: () => {},
  on: () => {},
  off: () => {},
} as any;

const Template: ComponentStory<typeof JoinRoomForm> = (args) => (
  <SocketContext.Provider value={dummySocket}>
    <JoinRoomForm />
  </SocketContext.Provider>
);

export const Default = Template.bind({});
