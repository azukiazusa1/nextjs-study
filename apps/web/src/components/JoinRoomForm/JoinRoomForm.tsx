import { REQ_EVENTS, RES_EVENTS } from 'models/session';
import { useRouter } from 'next/router';
import React from 'react';
import { useContext, useEffect, useRef, useState } from 'react';

import { SocketContext } from '@/context/socket';

import Avatar from '../Avatar/Avatar';

const avatarAltTexts = new Map([
  [1, '髭をはやしてオレンジ色の服を着ている男性'],
  [2, 'ショートヘアで赤い服を着ている女性'],
  [3, 'ちょび髭を生やした赤と黒のチェックシャツの男性'],
  [4, 'ミディアムヘアで緑の服を着ている女性'],
  [5, 'フラットデザインのキツネ'],
  [6, 'フラットデザインのペンギン'],
  [7, 'フラットデザインのアザラシ'],
  [8, 'フラットデザインのタコ'],
]);

const JoinRoomForm = () => {
  const socket = useContext(SocketContext);
  const router = useRouter();
  const [defaultUsername, setDefaultUsername] = useState('');
  const [avatar, setAvatar] = useState('1');
  const [hasError, setHasError] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);

  const selected = 'ring w-12 h-12 rounded-full ring-main';

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const username = inputRef.current?.value;
    if (!username) {
      setHasError(true);
      return;
    }
    setHasError(false);

    const score = Number(localStorage.getItem('score') || 0);

    socket.emit(REQ_EVENTS.JOIN_ROOM, {
      score,
      username,
      avatar,
    });

    localStorage.setItem('username', username);
    localStorage.setItem('avatar', avatar);
  };

  useEffect(() => {
    socket.on(RES_EVENTS.JOINED, ({ roomId }: { roomId: string }) => {
      router.push(`/room/${roomId}`);
    });

    setDefaultUsername(localStorage.getItem('username') || '');
    setAvatar(localStorage.getItem('avatar') || '1');

    return () => {
      socket.off(RES_EVENTS.JOINED);
    };
  }, [router, socket]);

  return (
    <div className="card flex-shrink-0 w-full max-w-sm shadow-2xl bg-base-100">
      <div className="card-body">
        <form onSubmit={handleSubmit}>
          <div className="form-control">
            <label className="label" htmlFor="username">
              <span className="label-text">名前</span>
            </label>
            <input
              type="text"
              id="username"
              defaultValue={defaultUsername}
              ref={inputRef}
              className={`input input-bordered ${hasError ? 'border-red-500' : ''}`}
            />
            {hasError && (
              <p role="alert" className="font-medium text-red-500 text-xs mt-2">
                名前を入力してください
              </p>
            )}
          </div>
          <div className="grid mt-2 grid-cols-4 gap-4 place-items-center">
            {[...Array(8).keys()]
              .map((i) => ++i)
              .map((i) => (
                <button
                  type="button"
                  key={i}
                  className={`${avatar === String(i) && selected}`}
                  onClick={() => setAvatar(String(i))}
                >
                  <Avatar src={`/images/${i}.png`} alt={avatarAltTexts.get(i) || ''} />
                </button>
              ))}
          </div>
          <div className="form-control mt-6">
            <button className="btn btn-primary">JOIN👏</button>
          </div>
        </form>
      </div>
    </div>
  );
};

export default JoinRoomForm;
