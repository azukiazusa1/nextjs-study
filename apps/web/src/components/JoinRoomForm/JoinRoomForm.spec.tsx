import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { REQ_EVENTS } from 'models';
import React from 'react';

import { SocketContext } from '@/context/socket';

import JoinRoomForm from './JoinRoomForm';

describe('hooks/useCountdown', () => {
  const mockSocket: any = {
    emit: jest.fn(),
    on: jest.fn(),
    off: jest.fn(),
  } as any;

  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <SocketContext.Provider value={mockSocket}>{children}</SocketContext.Provider>
  );

  beforeEach(() => {
    localStorage.clear();
    mockSocket.emit.mockReset();
  });

  test('名前を入力してサーバーへ入室リクエスト送信すること', async () => {
    render(<JoinRoomForm />, {
      wrapper,
    });

    await userEvent.type(screen.getByLabelText('名前'), 'test');
    await userEvent.click(screen.getByRole('button', { name: 'JOIN👏' }));

    expect(mockSocket.emit).toHaveBeenCalledWith(REQ_EVENTS.JOIN_ROOM, {
      username: 'test',
      avatar: '1',
      score: 0,
    });
  });

  it('should render correctly', () => {
    const { container } = render(<JoinRoomForm />, {
      wrapper,
    });
    expect(container).toMatchSnapshot();
  });

  test('名前を入力していない場合、入室リクエストを送信せずエラーメッセージを表示する', async () => {
    render(<JoinRoomForm />, {
      wrapper,
    });

    await userEvent.click(screen.getByRole('button', { name: 'JOIN👏' }));

    expect(mockSocket.emit).not.toHaveBeenCalled();
    expect(screen.getByRole('alert')).toHaveTextContent('名前を入力してください');
  });

  test('アバターをクリックして選択する', async () => {
    render(<JoinRoomForm />, {
      wrapper,
    });

    await userEvent.click(screen.getByAltText('フラットデザインのキツネ'));
    await userEvent.type(screen.getByLabelText('名前'), 'test');
    await userEvent.click(screen.getByRole('button', { name: 'JOIN👏' }));

    expect(mockSocket.emit).toHaveBeenCalledWith(REQ_EVENTS.JOIN_ROOM, {
      username: 'test',
      avatar: '5',
      score: 0,
    });
  });

  test('サーバーへリクエスト後 localStorage に保存する', async () => {
    render(<JoinRoomForm />, {
      wrapper,
    });

    await userEvent.click(screen.getByAltText('フラットデザインのキツネ'));
    await userEvent.type(screen.getByLabelText('名前'), 'test');
    await userEvent.click(screen.getByRole('button', { name: 'JOIN👏' }));

    expect(mockSocket.emit).toHaveBeenCalledWith(REQ_EVENTS.JOIN_ROOM, {
      username: 'test',
      avatar: '5',
      score: 0,
    });

    expect(localStorage.getItem('username')).toBe('test');
    expect(localStorage.getItem('avatar')).toBe('5');
  });

  test('初期値に localStorage の値を使用する', async () => {
    localStorage.setItem('username', 'username');
    localStorage.setItem('avatar', '3');
    localStorage.setItem('score', '20');

    render(<JoinRoomForm />, {
      wrapper,
    });

    await userEvent.click(screen.getByRole('button', { name: 'JOIN👏' }));

    expect(mockSocket.emit).toHaveBeenCalledWith(REQ_EVENTS.JOIN_ROOM, {
      username: 'username',
      avatar: '3',
      score: 20,
    });
  });

  test('ボタンをクリックした時、loading が表示される', async () => {
    render(<JoinRoomForm />, {
      wrapper,
    });

    await userEvent.type(screen.getByLabelText('名前'), 'test');
    await userEvent.click(screen.getByRole('button', { name: 'JOIN👏' }));

    expect(screen.getByRole('alert')).toHaveTextContent('loading');
  });
});
