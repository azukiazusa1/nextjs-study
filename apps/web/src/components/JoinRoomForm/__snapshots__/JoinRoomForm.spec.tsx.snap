// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooks/useCountdown should render correctly 1`] = `
<div>
  <div
    class="card flex-shrink-0 w-full max-w-sm shadow-2xl bg-base-100"
  >
    <div
      class="card-body"
    >
      <form>
        <div
          class="form-control"
        >
          <label
            class="label"
            for="username"
          >
            <span
              class="label-text"
            >
              名前
            </span>
          </label>
          <input
            class="input input-bordered "
            id="username"
            type="text"
            value=""
          />
        </div>
        <div
          class="grid mt-2 grid-cols-4 gap-4 place-items-center"
        >
          <button
            class="ring w-12 h-12 rounded-full ring-main"
            type="button"
          >
            <div
              class="avatar  undefined"
            >
              <div
                class="w-12 h-12 rounded-full"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2748%27%20height=%2748%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="髭をはやしてオレンジ色の服を着ている男性"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
            </div>
          </button>
          <button
            class="false"
            type="button"
          >
            <div
              class="avatar  undefined"
            >
              <div
                class="w-12 h-12 rounded-full"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2748%27%20height=%2748%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="ショートヘアで赤い服を着ている女性"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
            </div>
          </button>
          <button
            class="false"
            type="button"
          >
            <div
              class="avatar  undefined"
            >
              <div
                class="w-12 h-12 rounded-full"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2748%27%20height=%2748%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="ちょび髭を生やした赤と黒のチェックシャツの男性"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
            </div>
          </button>
          <button
            class="false"
            type="button"
          >
            <div
              class="avatar  undefined"
            >
              <div
                class="w-12 h-12 rounded-full"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2748%27%20height=%2748%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="ミディアムヘアで緑の服を着ている女性"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
            </div>
          </button>
          <button
            class="false"
            type="button"
          >
            <div
              class="avatar  undefined"
            >
              <div
                class="w-12 h-12 rounded-full"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2748%27%20height=%2748%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="フラットデザインのキツネ"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
            </div>
          </button>
          <button
            class="false"
            type="button"
          >
            <div
              class="avatar  undefined"
            >
              <div
                class="w-12 h-12 rounded-full"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2748%27%20height=%2748%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="フラットデザインのペンギン"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
            </div>
          </button>
          <button
            class="false"
            type="button"
          >
            <div
              class="avatar  undefined"
            >
              <div
                class="w-12 h-12 rounded-full"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2748%27%20height=%2748%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="フラットデザインのアザラシ"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
            </div>
          </button>
          <button
            class="false"
            type="button"
          >
            <div
              class="avatar  undefined"
            >
              <div
                class="w-12 h-12 rounded-full"
              >
                <span
                  style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
                >
                  <span
                    style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
                  >
                    <img
                      alt=""
                      aria-hidden="true"
                      src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2748%27%20height=%2748%27/%3e"
                      style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                    />
                  </span>
                  <img
                    alt="フラットデザインのタコ"
                    data-nimg="intrinsic"
                    decoding="async"
                    src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                    style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
                  />
                  <noscript />
                </span>
              </div>
            </div>
          </button>
        </div>
        <div
          class="form-control mt-6"
        >
          <button
            class="btn btn-primary"
          >
            JOIN👏
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
